src/authentication.cpp

```diff
@@ -0,0 +1,239 @@
+            result << val << '\n';
+        }
+        result << "\n";
+
+        sorted_fields.clear();
+
+        // and the signed header list
+
+        for (const auto& hd : signed_headers) {
+            sorted_fields.push_back(hd);
+        }
+        std::sort(sorted_fields.begin(), sorted_fields.end());
+        {
+            bool first = true;
+            for (const auto& i : sorted_fields) {
+                result << (first ? "" : ";") << i;
+                first = false;
+            }
+            result.copyfmt(state);
+            result << '\n';
+        }
+
+        //and the payload signature
+
+        if (auto req = request.get().find("X-Amz-Content-SHA256"); req != request.get().end()) {
+            result << req->value();
+        }
+        else {
+            result << "UNSIGNED-PAYLOAD";
+        }
+
+        return result.str();
+    }
+
+    std::string string_to_sign(
+        const static_buffer_request_parser& request,
+        const boost::urls::url_view& url,
+        const std::string_view& date,
+        const std::string_view& region,
+        const std::string_view& canonical_request)
+    {
+        std::stringstream result;
+        result << "AWS4-HMAC-SHA256\n";
+        result << request.get().at("X-Amz-Date") << '\n';
+        result << date << '/' << region << "/s3/aws4_request\n";
+        result << irods::s3::authentication::hex_encode(irods::s3::authentication::hash_sha_256(canonical_request));
+        return result.str();
+    }
+} //namespace
+bool irods::s3::authentication::authenticates(
+    rcComm_t& conn,
+    const static_buffer_request_parser& request,
+    const boost::urls::url_view& url)
+{
+    std::vector<std::string> auth_fields, credential_fields, signed_headers;
+    // should be equal to something like
+    // [ 'AWS4-SHA256-HMAC Credential=...', 'SignedHeaders=...', 'Signature=...']
+    //
+    boost::split(auth_fields, request.get().at("Authorization"), boost::is_any_of(","));
+
+    // Strip the names and such
+    for (auto& field : auth_fields) {
+        field = field.substr(field.find('=') + 1);
+    }
+
+    // Break up the credential field.
+    boost::split(credential_fields, auth_fields[0], boost::is_any_of("/"));
+
+    auto& access_key_id = credential_fields[0];
+    auto& date = credential_fields[1];
+    auto& region = credential_fields[2];
+    // Look, covering my bases here seems prudent.
+    auto& aws_service = credential_fields[3];
+
+    auto& signature = auth_fields[2];
+
+    boost::split(signed_headers, auth_fields[1], boost::is_any_of(";"));
+
+    auto canonical_request = canonicalize_request(request, url, signed_headers);
+    std::cout << "Canon request==========================" << std::endl;
+    std::cout << canonical_request << std::endl;
+    std::cout << "=======================================" << std::endl;
+
+    auto sts = string_to_sign(request, url, date, region, canonical_request);
+
+    std::cout << sts << std::endl;
```

Comment:
Prefer `\n` instead of `std::endl`.

Each call to `std::endl` makes a call to `std::flush` which hurts performance. Use `std::endl` only when you need to.

Of course, this goes away once a logger is added to the implementation.